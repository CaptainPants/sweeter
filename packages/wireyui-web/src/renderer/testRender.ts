import { WebRenderer } from './WebRenderer.js';

export interface TestRenderResult {
    readonly nodes: NodeListOf<ChildNode>;
    readonly getHTML: () => string;
    readonly dispose: () => void;
}

/**
 * Convenience function for rendering isolated JSX chunks for unit tests
 * @param callback
 * @returns
 */
export function testRender(callback: () => JSX.Element): TestRenderResult {
    const renderer = new WebRenderer();

    const root = document.createElement('div');

    const dispose = renderer.start(root, callback);

    return {
        nodes: root.childNodes,
        getHTML: () => root.innerHTML,
        dispose,
    };
}
