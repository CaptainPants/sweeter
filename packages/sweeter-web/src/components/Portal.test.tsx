/* @jsxImportSource .. */

import {
    $mutable,
    type ComponentInit,
    Dynamic,
    Portal,
} from '@captainpants/sweeter-core';
import { testRender } from '../test/testRender.js';
import {
    GlobalCssStylesheet,
    IncludeStylesheet,
    stylesheet,
} from '../index.js';

// it('Portal element content is added', () => {
//     const target = document.createElement('div');

//     const { dispose, getHTML } = testRender(() => (
//         <Portal target={target}>
//             <div>CONTENT</div>
//         </Portal>
//     ));

//     // The contents of our in-tree render should be empty, but will have hook placeholders
//     expect(getHTML()).toMatchSnapshot();

//     expect(target.outerHTML).toMatchSnapshot();

//     dispose();

//     // Should have been cleared
//     expect(target.outerHTML).toMatchSnapshot();
// });

it('Portal elements unmount correctly', () => {
    const shouldMount = $mutable(true);

    const style = new GlobalCssStylesheet({
        id: 'style',
        content: stylesheet`
            div {
                color: red;
            }
        `,
    });

    let mounted = false;

    // eslint-disable-next-line @typescript-eslint/ban-types
    function TestComponent(props: {}, init: ComponentInit) {
        init.onMount(() => {
            mounted = true;

            return () => {
                mounted = false;
            };
        });

        return <></>;
    }

    const target = document.createElement('div');
    target.id = 'host';
    document.body.append(target);

    const { dispose } = testRender(() => (
        <Dynamic<boolean> value={shouldMount}>
            {(visible) =>
                visible ? (
                    <>
                        MOUNTED
                        <Portal target={target}>
                            <div>MOUNTED</div>
                            <IncludeStylesheet stylesheet={style} />
                        </Portal>
                        <TestComponent />
                    </>
                ) : (
                    'NOT MOUNTED'
                )
            }
        </Dynamic>
    ));

    shouldMount.value = true;
    expect(mounted).toStrictEqual(true);

    expect(document.head.outerHTML).toMatchSnapshot();
    expect(document.body.outerHTML).toMatchSnapshot();

    shouldMount.value = false;
    expect(mounted).toStrictEqual(false);

    // Should have been cleared
    expect(document.head.outerHTML).toMatchSnapshot();
    expect(document.body.outerHTML).toMatchSnapshot();

    shouldMount.value = true;
    expect(mounted).toStrictEqual(true);

    // Everything should be back
    expect(document.head.outerHTML).toMatchSnapshot();
    expect(document.body.outerHTML).toMatchSnapshot();

    dispose();

    // Everything should be cleared
    expect(mounted).toStrictEqual(false);

    expect(document.head.outerHTML).toMatchSnapshot();
    expect(document.body.outerHTML).toMatchSnapshot();
});
