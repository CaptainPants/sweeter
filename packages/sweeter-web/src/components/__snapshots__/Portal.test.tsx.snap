// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Portal elements unmount correctly (css class ref) 1`] = `
"<head><style data-id=\\"name\\">
.name div {
    color: red;
}
</style></head>"
`;

exports[`Portal elements unmount correctly (css class ref) 2`] = `"<body><div id=\\"host3\\"><div class=\\"name\\">MOUNTED</div></div><div>MOUNTED<!--Hooks(Portal): subscribeToChanges--><!--Hooks(TestComponent): onMount--></div></body>"`;

exports[`Portal elements unmount correctly (css class ref) 3`] = `"<head></head>"`;

exports[`Portal elements unmount correctly (css class ref) 4`] = `"<body><div id=\\"host3\\"></div><div>NOT MOUNTED</div></body>"`;

exports[`Portal elements unmount correctly (css class ref) 5`] = `
"<head><style data-id=\\"name\\">
.name div {
    color: red;
}
</style></head>"
`;

exports[`Portal elements unmount correctly (css class ref) 6`] = `"<body><div id=\\"host3\\"><div class=\\"name\\">MOUNTED</div></div><div>MOUNTED<!--Hooks(Portal): subscribeToChanges--><!--Hooks(TestComponent): onMount--></div></body>"`;

exports[`Portal elements unmount correctly (css class ref) 7`] = `"<head></head>"`;

exports[`Portal elements unmount correctly (css class ref) 8`] = `"<body><div id=\\"host3\\"></div></body>"`;
